qualtran/qir_example.ipynb

```diff
@@ -0,0 +1,685 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import pyqir\n",
+    "import qualtran\n",
+    "from qualtran.bloqs.basic_gates import Hadamard, XGate, YGate, ZGate, Rx, Ry, Rz, TGate, TwoBitSwap, CNOT, Toffoli\n",
+    "from qualtran.bloqs.qft.qft_text_book import QFTTextBook\n",
+    "import pyqir\n",
+    "from pyqir import (\n",
+    "    BasicBlock,\n",
+    "    Builder,\n",
+    "    Context,\n",
+    "    Function,\n",
+    "    Linkage,\n",
+    "    Module,\n",
+    "    ModuleFlagBehavior,\n",
+    ")"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 60,
+   "outputs": [],
+   "source": [
+    "import attrs\n",
+    "from typing import Dict\n",
+    "from qualtran import Bloq, BloqBuilder, SoquetT, Signature, Register\n",
+    "\n",
+    "\n",
+    "@attrs.frozen\n",
+    "class Swap(Bloq):\n",
+    "    n: int\n",
+    "\n",
+    "    @property\n",
+    "    def signature(self):\n",
+    "        return Signature.build(x=self.n, y=self.n)\n",
+    "\n",
+    "    def build_composite_bloq(\n",
+    "            self, bb: BloqBuilder, *, x: SoquetT, y: SoquetT\n",
+    "    ) -> Dict[str, SoquetT]:\n",
+    "        xs = bb.split(x)\n",
+    "        ys = bb.split(y)\n",
+    "\n",
+    "        for i in range(self.n):\n",
+    "            xs[i], ys[i] = bb.add(CNOT(), ctrl=xs[i], target=ys[i])\n",
+    "        return {\n",
+    "            'x': bb.join(xs),\n",
+    "            'y': bb.join(ys),\n",
+    "        }\n",
+    "\n",
+    "@attrs.frozen\n",
+    "class ExampleBaseBloq(Bloq):\n",
+    "    @property\n",
+    "    def signature(self):\n",
+    "        return Signature.build(x=1, y=1)\n",
+    "\n",
+    "@attrs.frozen\n",
+    "class ExampleHighLevelBloq(Bloq):\n",
+    "    n: int\n",
+    "\n",
+    "    @property\n",
+    "    def signature(self):\n",
+    "        return Signature.build(x=self.n, y=self.n)\n",
+    "\n",
+    "    def build_composite_bloq(\n",
+    "            self, bb: BloqBuilder, *, x: SoquetT, y: SoquetT\n",
+    "    ) -> Dict[str, SoquetT]:\n",
+    "        xs = bb.split(x)\n",
+    "        ys = bb.split(y)\n",
+    "\n",
+    "        for i in range(self.n):\n",
+    "            xs[i], ys[i] = bb.add(ExampleBaseBloq(), x=xs[i], y=ys[i])\n",
+    "        return {\n",
+    "            'x': bb.join(xs),\n",
+    "            'y': bb.join(ys),\n",
+    "        }\n",
+    "\n",
+    "@attrs.frozen\n",
+    "class ExampleNonTrivialShapeBloq(Bloq):\n",
```

Comment:
Another example to add (maybe as a followup) would be `ExampleNonTrivialSideBloq` which has LEFT and RIGHT directional registers instead of all registers being THRU registers. 

I think in terms of QIR, a LEFT register implies de-allocation happens inside that function and a `RIGHT` register corresponds to allocations happening inside that function. If the concepts don't translate directly (eg: because QIR assumes you take care of qubit allocations and deallocations globally and have a sea of globally allocated qubits) then I think both LEFT & RIGHT registers would be treated as inputs to the corresponding the QIR function (we do this when supporting cirq style decomposition where `decompose_from_registers()` method gets both LEFT & RIGHT registers as preallocated ancilla qubits.

Tl;Dr - Converting bloqs with directional registers should be tracked as a future improvement in the issue